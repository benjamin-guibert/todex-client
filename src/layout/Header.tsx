import React, { FC, useContext } from 'react'
import { InitializationStatus, MetaMaskContext } from './MetaMaskContext'
import { ethers } from 'ethers'
import { printAmount, shortenAddress } from 'libraries/helpers'
import { ExchangeContext } from './ExchangeContext'
import Container from 'react-bootstrap/Container'
import Navbar from 'react-bootstrap/Navbar'
import Button from 'react-bootstrap/Button'
import StyledHeader from './Header.style'

const printBalance = (balance?: string) => {
  return printAmount(balance || null)
}

const Header: FC = () => {
  const { initializationStatus, account, connect } = useContext(MetaMaskContext)
  const { ethBalance, tokenBalance } = useContext(ExchangeContext)

  return (
    <StyledHeader bg="dark" variant="dark">
      <Container fluid>
        <Navbar.Brand>ToDEX</Navbar.Brand>
        {initializationStatus == InitializationStatus.Initialized && (
          <>
            {!!ethBalance && <Navbar.Text>{`${ethers.constants.EtherSymbol} ${printBalance(ethBalance)}`}</Navbar.Text>}
            {!!tokenBalance && <Navbar.Text>{`ToDEX ${printBalance(tokenBalance)}`}</Navbar.Text>}
            {!!account && <Navbar.Text>{shortenAddress(account)}</Navbar.Text>}
            {!account && <Button onClick={connect}>Connect Wallet</Button>}
          </>
        )}
      </Container>
    </StyledHeader>
  )
}

export default Header
