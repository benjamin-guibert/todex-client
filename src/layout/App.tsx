import React, { FC, useEffect, useRef, useState } from 'react'
import { Web3Provider } from '@ethersproject/providers'
import { Token } from '../contracts/Token'
import { MetaMaskContext, useMetaMaskContext } from './MetaMaskContext'
import { ExchangeContext, useExchangeContext } from './ExchangeContext'
import Alert from 'react-bootstrap/Alert'
import Header from './Header'
import MetaMaskModal from './MetaMaskModal'
import { initializeContract } from 'libraries/contracts/token'
import Forms from './Forms'
import TradeHistory from './TradeHistory'
import Orders from './Orders'
import StyledApp, { MainContainer } from './App.style'

const App: FC = () => {
  const metaMaskContextValue = useMetaMaskContext()
  const exchangeContextValue = useExchangeContext()
  const { initialize: initializeMetaMask, getAccount, account } = metaMaskContextValue
  const { initialized, initialize: initializeExchange, setAccount } = exchangeContextValue
  const initializeMetaMaskRef = useRef<() => Promise<Web3Provider | undefined>>(initializeMetaMask)
  const getAccountRef = useRef<() => Promise<string | null | undefined>>(getAccount)
  const initializeExchangeRef = useRef<(provider: Web3Provider, token: Token) => Promise<boolean>>(initializeExchange)
  const setExchangeAccountRef = useRef<(account: string | null | undefined) => void>(setAccount)
  const [error, setError] = useState<string | null>(null)

  useEffect(() => {
    const initialize = async (): Promise<void> => {
      const provider = await initializeMetaMaskRef.current()
      if (!provider) {
        setError('MetaMask not initialized.')
        return
      }

      let token: Token

      try {
        token = initializeContract(provider).token
      } catch {
        setError('Error while initializing token.')
        return
      }

      const exchangeInitialized = await initializeExchangeRef.current(provider, token)
      if (!exchangeInitialized) {
        return setError('Error while initializing exchange.')
      }

      getAccountRef.current()
    }

    initialize()
  }, [initializeMetaMaskRef, initializeExchangeRef, getAccountRef])

  useEffect(() => {
    setExchangeAccountRef.current(account)
  }, [account])

  return (
    <StyledApp className="bg-dark text-light">
      <MetaMaskContext.Provider value={metaMaskContextValue}>
        <ExchangeContext.Provider value={exchangeContextValue}>
          <Header />
          {!!error && <Alert variant="danger">{error}</Alert>}
          <MainContainer fluid>
            {initialized && (
              <>
                <Forms />
                <Orders />
                <TradeHistory />
              </>
            )}
          </MainContainer>
          <MetaMaskModal />
        </ExchangeContext.Provider>
      </MetaMaskContext.Provider>
    </StyledApp>
  )
}

export default App
