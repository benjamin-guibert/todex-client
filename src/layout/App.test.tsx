import * as React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import App from './App'
import Header from './Header'

describe('<App />', () => {
  it('should render', () => {
    ReactDOM.createPortal = jest.fn((element) => {
      return element
    }) as unknown as () => React.ReactPortal
    const component = renderer.create(<App />)

    expect(component.root.findByType(Header)).toBeDefined()
    expect(component.toJSON()).toMatchSnapshot()
  })
})
