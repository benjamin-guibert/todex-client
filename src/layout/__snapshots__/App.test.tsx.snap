// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should render when not initialized 1`] = `
<div
  className="sc-gKclnd bvMvZo bg-dark text-light"
>
  <nav
    className="sc-bdvvtL dqBChi navbar navbar-expand navbar-light bg-light"
  >
    <div
      className="container-fluid"
    >
      <span
        className="navbar-brand"
      >
        ToDEX
      </span>
    </div>
  </nav>
  <div
    className="sc-jRQBWg ccJCcb container-fluid"
  />
  <div
    className="fade modal-backdrop"
    onClick={[Function]}
  />
  <div
    aria-modal={true}
    className="fade modal"
    onClick={[Function]}
    onMouseUp={[Function]}
    role="dialog"
    style={Object {}}
    tabIndex={-1}
  >
    <div
      className="modal-dialog modal-sm modal-dialog-centered"
      onMouseDown={[Function]}
    >
      <div
        className="modal-content"
      >
        <div
          className="modal-header"
        >
          <div
            className="modal-title h4"
          >
            MetaMask
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
